### DISCUSSION TITLE
[RFC] Spinners and user scripts

### DISCUSSION THREAD

**Discussion by jakobrosenberg** [pycom/pymakr-vsc#259] [Ideas]
In Pymakr we use spinners to show that a device isn't idle. This works as intended for temporary jobs but for long or inifinte scripts it may leave the user waiting for the spinner to go away. Ideally the interface would behave the way the most users expect it to behave.

I propose we keep the spinner for Pymakr actions and use a different icon for running user scripts.

Example of how we could do this:
**Idle:** terminal icon
**Pymakr action:** animated spinner
**User script:** lightning

**Comment by VirtualWolf** [comment]
I like it. Would there still be the concept of connecting/disconnecting a device? I guess that'd have to shift from the lightning bolt if so...

**Reply by jakobrosenberg** [reply]
If I remember correctly, VSCode allows for one icon per item, so the icon would have to cover every state:
- offline,
- idle (terminal)
- running (user script)
- busy (Pymakr action)

**Reply by jakobrosenberg** [reply]
Yes. We need to be able to disconnect devices. I don't know if we can use the existing connect/disconnect icons. Think we should change them?

**Reply by Josverl** [reply]
I would also like an option to cancel such a waiting state, in case the state in pymakr does not reflect the actual state.
This can happen when 
 - pymakr does not recognize the return / end-state
 - or when i pysically reset a device 
 - or when a script resets the device.

**Reply by Josverl** [reply]
Also , can we have a tooltip over the icon to have a short description of what the state is/icon means ?

**Reply by jakobrosenberg** [reply]
That's a great idea! We can't do it for the icon itself, but we can do it for the device, so that if you hover over it, it would show the current running action.

**Reply by VirtualWolf** [reply]
> I would also like an option to cancel such a waiting state, in case the state in pymakr does not reflect the actual state.

Yeah this'd be great as well. I've run into that sort of scenario a few times.

**Reply by jakobrosenberg** [reply]
Been doing some work on this today.
Current main state:
- disconnected
- busy
- idle

New main state:
- offline (not plugged in)
- disconnected (plugged in but not connected in Pymakr)
- idle
- action (Pymakr)
- script (User)

I expect the added clarity will make it easier to detect and solve hanging actions, even by force.

**Reply by jakobrosenberg** [reply]
Not sure if `offline` and `disconnected` should be swapped.

**Reply by jakobrosenberg** [reply]
![states](https://user-images.githubusercontent.com/4153004/176608974-bb30906a-d04a-4e8f-b9d7-2291659b409e.gif)
